### ISSUE TITLE
2019-12-12 tc113 M3SD doesn't execute with MRT related parameters but only with a subset

### ISSUE THREAD

**Issue by tensfeldt** [tensfeldt/openNCA#108]
This test case for M3SD failed during the M1 E2E evaluation of the openNCA application when parameterset=="PARAMETERLIST" but could execute only after MRT related parameters were removed.

In this issue, this is illustrated by using an updated parameter list but the same behavior is illustated. The script below illustrates the method of replicating the issue. Note that the error message indicates that DOF is non-numeric but testing that directly from the input concentration dataset doesn't confirm that error message from the computation engine.

The files for the test case tc113 are loaded to sharepoint/onedrive. 

## FAIL: parameterset=="PARAMETERLIST" using a more recent edition of the MCT than what was used in the test case
thus steps to replicate includes a)  reading the original MCT b) replacing PARAMETERLIST/PARAMETERDISPLAYLIST with the values in the latest/updated MCT.

### replace parameterlist/parameterdisplaylist
```
# Execute with parameterlist used with 2019-12-06_M3SD MCT 
mct$PARAMETERLIST <- M3SD_20191206_PARAMETERLIST
mct$PARAMETERDISPLAYLIST <- M3SD_20191206_PARAMETERDISPLAYLIST
```

```
> parameterset <- "PARAMETERLIST"
> mct$PARAMETERLIST
[1] "AUCALL;AUCALLDN;AUCDN;AUCINFO;AUCINFOC;AUCINFODN;AUCINFP;AUCINFPC;AUCINFPDN;AUCLAST;AUCLASTC;AUCLASTDN;AUCT;AUCT1_T2;AUCTDN;AUCXPCTO;AUCXPCTP;AUMCINFO;AUMCINFP;AUMCLAST;AUMCXPTO;AUMCXPTP;CENDINF;CENDINFDN;CEST;CLAST;CLO;CLOW;CLP;CLPW;CMAX;CMAXDN;CMIN;CMINDN;DOFi;DOSE;DOSEC;KEL;KELC0;KELNOPT;KELR;KELRSQ;KELRSQA;KELTMHI;KELTMLO;LASTTIME;MRTIVIFO;MRTIVIFP;MRTLAST;TENDINF;THALF;THALFF;TLAST;TMAX;TMIN;VSSP;VSSPW;VSSO;VSSOW;VZO;VZOW;VZP;VZPW;FLGACCEPTPREDOSE;FLGACCEPTTMAX;FLGACCEPTKEL"
> # parameters removed from mct$PARAMETERLIST that executed successfully
> setdiff(unlist(strsplit(mct$PARAMETERLIST, ";")), unlist(strsplit(mct.saved$PARAMETERLIST, ";")))
 [1] "AUCALL"    "AUCALLDN"  "AUCDN"     "AUCINFO"   "AUCINFOC"  "AUCINFODN" "AUCINFPC"  "AUCINFPDN" "AUCLASTC"  "AUCLASTDN" "AUCT"      "AUCT1_T2"  "AUCTDN"    "AUCXPCTO"  "AUMCINFO" 
[16] "AUMCXPTO"  "AUMCXPTP"  "CENDINF"   "CENDINFDN" "CEST"      "CLO"       "CLOW"      "CLPW"      "CMAXDN"    "CMINDN"    "DOFi"      "DOSEC"     "MRTIVIFO"  "MRTIVIFP"  "MRTLAST"  
[31] "TENDINF"   "VSSP"      "VSSPW"     "VSSO"      "VSSOW"     "VZO"       "VZOW"      "VZP"       "VZPW"     
> 
> (tclabel <- paste0("results.", testcase.id, ".", mct$MODEL, ".", mct$DOSINGTYPE, ".", mct$AUCMETHOD))
[1] "results.tc113.M3.SD.LINLOG"
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in mrt_last(conc = tmp_df[, map_data$CONC], time = tmp_df[, map_data$TIME], : Error in mrt_last: dof is not a numeric value
For SDEID 49420991
In addition: Warning messages:
1: In validate_timeconc_data(map, data, verbose = verbose) :
  No tau information provided in 'map'.
2: In validate_timeconc_data(map, data, verbose = verbose) :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in mrt_last(conc = tmp_df[, map_data$CONC], time = tmp_df[, map_data$TIME], : Error in mrt_last: dof is not a numeric value
For SDEID 49420991 > r <- results_list$data_out
> e <- results_list$est_data
> f <- results_list$flag_data
> x <- unlist(strsplit(mct$PARAMETERDISPLAYLIST, ";"))
> y <- unlist(strsplit(mct$PARAMETERLIST, ";"))
> z <- names(r)
> setdiff(x, z)
[1] "PERIODU"              "PERIOD"               "FLGACCEPTKELCRIT"     "FLGACCEPTPREDOSECRIT" "SPANRATIOCRIT"       
> setdiff(y, z)
 [1] "AUCALL"    "AUCALLDN"  "AUCDN"     "AUCINFO"   "AUCINFOC"  "AUCINFODN" "AUCINFPC"  "AUCINFPDN" "AUCLASTC"  "AUCLASTDN" "AUCT"      "AUCT1_T2"  "AUCTDN"    "AUCXPCTO"  "AUMCINFO" 
[16] "AUMCINFP"  "AUMCLAST"  "AUMCXPTO"  "AUMCXPTP"  "CENDINF"   "CENDINFDN" "CEST"      "CLAST"     "CLO"       "CLOW"      "CLPW"      "CMAXDN"    "CMIN"      "CMINDN"    "DOFi"     
[31] "DOSEC"     "KELC0"     "KELR"      "KELRSQA"   "LASTTIME"  "MRTIVIFO"  "MRTIVIFP"  "MRTLAST"   "TENDINF"   "THALFF"    "TMIN"      "VSSP"      "VSSPW"     "VSSO"      "VSSOW"    
[46] "VZO"       "VZOW"      "VZPW"     
> 
> # Check if DOF is numeric
> is.numeric(d$DOF)
[1] TRUE
```

## SUCCESS: parameterset=="PARAMETERDISPLAYLIST" using a more recent edition of the MCT than what was used in the test case
PARAMETERDISPLAYLIST has a very limited subset of the PARAMETERLIST parameters.

```
> # Execute with parameterlist used with AN13604 successfully
> parameterset <- "PARAMETERDISPLAYLIST"
> mct$PARAMETERDISPLAYLIST
[1] "SUBJID;TREATXT;PKBDFLD;PKTERM;PERIODU;PERIOD;VISITU;VISIT;DOSE;DOSEU;ACTTRT;AUCINFP;AUCLAST;AUCXPCTP;CLP;CMAX;KEL;KELNOPT;KELRSQ;KELTMHI;KELTMLO;THALF;TLAST;TMAX;VZP;FLGACCEPTKEL;FLGACCEPTKELCRIT;FLGACCEPTPREDOSE;FLGACCEPTPREDOSECRIT;FLGACCEPTTMAX;SPANRATIOCRIT"
> # parameters removed from mct$PARAMETERDISPLAYLIST that executed successfully
> setdiff(unlist(strsplit(mct$PARAMETERDISPLAYLIST, ";")), unlist(strsplit(mct.saved$PARAMETERDISPLAYLIST, ";")))
 [1] "SUBJID"               "TREATXT"              "PKBDFLD"              "PKTERM"               "PERIODU"              "PERIOD"               "VISITU"               "VISIT"               
 [9] "DOSEU"                "ACTTRT"               "TLAST"                "VZP"                  "FLGACCEPTKELCRIT"     "FLGACCEPTPREDOSECRIT" "SPANRATIOCRIT"       
> 
> (tclabel <- paste0("results.", testcase.id, ".", mct$MODEL, ".", mct$DOSINGTYPE, ".", mct$AUCMETHOD))
[1] "results.tc113.M3.SD.LINLOG"
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
unit_conversion : Unit Class 1 (Time) time_col:  TMAX TLAST KELTMLO KELTMHI THALF  parameters are scaled from  HR  to  HR  via scaling factor:  1 
unit_conversion : Unit Class 3 (Dose) dose_col:  DOSE  parameters are scaled from  UG  to  MG  via scaling factor:  0.001 
unit_conversion : Unit Class 4 (Volume) volume_col:  VZP  parameters are scaled from  ML  to  L  via scaling factor:  0.001 
unit_conversion : Unit Class 5 (Amount/Volume) conc_col:  CMAX  parameters are scaled from  NG/ML  to  NG/ML  via scaling factor:  1 
unit_conversion : Unit Class 6 (1/Time) kel_col:  KEL  parameters are scaled from  1/HR  to  1/HR  via scaling factor:  1 
unit_conversion : Unit Class 7 (Volume/Time) cl_col:  CLP  parameters are scaled from  ML/HR  to  ML/HR  via scaling factor:  1 
unit_conversion : Unit Class 8: (Amount.Time/Volume) auc_col:  AUCLAST AUCINFP  parameters are scaled from  NG.HR/ML  to  NG.HR/ML  via scaling factor:  1 
Warning messages:
1: In validate_timeconc_data(map, data, verbose = verbose) :
  No tau information provided in 'map'.
2: In validate_timeconc_data(map, data, verbose = verbose) :
  No told information provided in 'map'.
3: In run_M3_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'SPANRATIOCRIT' is not present in the dataset
4: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'AMOUNTU' #2# is not present in the dataset provided via 'map'
5: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'PKPTMU', 'CONCU' and 'DOSEU' #8# are not present in the dataset provided via 'map'
6: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'PKPTMU' and 'CONCU' #10# are not present in the dataset provided via 'map'
7: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU' #24# is not present in the dataset provided via 'map'
8: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU', 'PKPTMU and 'NORMBSU' #19# are not present in the dataset provided via 'map'
> r <- results_list$data_out
> e <- results_list$est_data
> f <- results_list$flag_data
> x <- unlist(strsplit(mct$PARAMETERDISPLAYLIST, ";"))
> y <- unlist(strsplit(mct$PARAMETERLIST, ";"))
> z <- names(r)
> setdiff(x, z)
[1] "PERIODU"              "PERIOD"               "FLGACCEPTKELCRIT"     "FLGACCEPTPREDOSECRIT" "SPANRATIOCRIT"       
> setdiff(y, z)
 [1] "AUCALL"    "AUCALLDN"  "AUCDN"     "AUCINFO"   "AUCINFOC"  "AUCINFODN" "AUCINFPC"  "AUCINFPDN" "AUCLASTC"  "AUCLASTDN" "AUCT"      "AUCT1_T2"  "AUCTDN"    "AUCXPCTO"  "AUMCINFO" 
[16] "AUMCINFP"  "AUMCLAST"  "AUMCXPTO"  "AUMCXPTP"  "CENDINF"   "CENDINFDN" "CEST"      "CLAST"     "CLO"       "CLOW"      "CLPW"      "CMAXDN"    "CMIN"      "CMINDN"    "DOFi"     
[31] "DOSEC"     "KELC0"     "KELR"      "KELRSQA"   "LASTTIME"  "MRTIVIFO"  "MRTIVIFP"  "MRTLAST"   "TENDINF"   "THALFF"    "TMIN"      "VSSP"      "VSSPW"     "VSSO"      "VSSOW"    
[46] "VZO"       "VZOW"      "VZPW" 
```

**Comment by jhhughes256** [comment]
As of 00cdcd3393 this issue still occurs.

**Comment by rudraya** [comment]
The parameter MRT is not being computed because the MCT contains "DOF", the openNCA code currently accounts "DOF1", please let me know if this needs to be changed.

**Comment by tensfeldt** [comment]
This issue tc113 M3SD was retested with commit 560bcad. The original issue has been cleared. However CMINDN, when specified in the PARAMETERLIST, is not generated. This issue will be closed and a separate issue will be opened specific to that defect.
